core/Symbols.cc

```diff
@@ -454,9 +456,16 @@ string TypeMemberRef::show(const GlobalState &gs) const {
     if (sym->name == core::Names::Constants::AttachedClass()) {
         auto attached = sym->owner.asClassOrModuleRef().data(gs)->attachedClass(gs);
         ENFORCE(attached.exists());
-        return fmt::format("T.attached_class (of {})", attached.show(gs));
+        // TODO: This change is needed for RBI generation.
+        return "T.attached_class";
+        // return fmt::format("T.attached_class (of {})", attached.show(gs));
     }
-    return showInternal(gs, sym->owner, sym->name, COLON_SEPARATOR);
+    auto owner = sym->owner;
+    // Don't show T.class_of(Foo)::Field; show Foo::Field.
+    if (sym->owner.asClassOrModuleRef().data(gs)->isSingletonClass(gs)) {
+        owner = sym->owner.asClassOrModuleRef().data(gs)->attachedClass(gs);
+    }
+    return showInternal(gs, owner, sym->name, COLON_SEPARATOR);
```

Comment:
there's a subtle distinction here

```ruby
class Foo
  class << self
    Field = 42
  end
end

p Foo::Feld # => NameError
```

I'd love to see where this comes up specifically, because there might be a better way to detect the case(s) we're trying to handle here.

Another one:

```ruby
class A
  extend T::Generic

  MyTemplate = type_template
end

p A::MyTemplate
```

In this case, Sorbet would print it as `T.class_of(A)::MyTemplate` even though `A::MyTemplate` would resolve. But:

```ruby
class A
  extend T::Generic

  class << self
    MyTemplate = type_member
  end
end

p A::MyTemplate # => NameError
```

This would result in the same globalstate, but different runtime behavior.

(That's kind of why I want to learn more about what the actual code that informed this change looks like.)